### PR TITLE
piston extruder 5339

### PR DESCRIPTION
**Pull request by BillStruve**
##Why (short version)
Printing by extruding a paste with a stepper motor driven piston does not require
temperature control of the extruder head or the bed.

It does require control of a fan to dry the paste after extrusion.
When not dried, the print will collapse when it is tall.
The manufacturer provides a warm air blower fan for this.
	M35 turns fan on, M36 turns fan off.
	
It also requires a means of adjusting the extrude rate during printing.
The manufacturer has provided a pot with a knob for the user.
	M37 A# - turns on/off the E multiplier Adjustment
	M37 S# - turns on/off Showing of the E multiplier
	M37 R##.# - adjusts the sensitivity of the knob (%)
		equation: y = mx + b  where y = multiplier,
		m = R/100/512,  x = ADC reading, b is calculated
		assuming (max x)/2 = 512 and at the midpoint of the
		adjustment, the multiplier will be 1.000
	
##What (overview)
Configuration.h
	add a section for a piston extruder printer
	add 2 defines to direct compiler
	add 2 default values
Marlin.h
	add 6 extern vars for use in Marlin_main.cpp,
		planner.cpp, and temperature.cpp
Marlin_main.cpp
	set initial values for vars
	add descriptions for M35, M36, and M37
	add code for M35, M36, and M37
	add case's for M35, M36, and M37
enum.h
	add a Prepare/Measure pair for ADC
pins_RAMPS.h
	define a pin for the dryer fan
	define a pin for the ADC input
planner.cpp
	alter extrusion with E multiplier
temperature.cpp
	setup for analog read
	use analog read to calculate E multiplier
	provide user feedback of E multiplier
example_configurations
	add MMM folder
	add a specific Configuration.h to MMM folder

### CODE REVIEW

File: Marlin/Configuration.h [hunk 1]
```diff
@@ -81,6 +81,22 @@
...
 
+//===========================================================================
+//======================== Piston Extruder Printer ==========================
+//===========================================================================
+// For a Piston Extruder printer replace configuration.h with the file in the
+// example_configurations/MMM directory.
+//
+
+//#define PISTON_EXTRUDER //used to set parameters for the Piston Extruder
+#ifdef PISTON_EXTRUDER
```
Comment (thinkyhead) [inline Marlin/Configuration.h]: Please use `#if ENABLED(...)` for options like these.

File: Marlin/Marlin_main.cpp [hunk 2]
```diff
@@ -4810,6 +4825,65 @@ inline void gcode_M31() {
 
 #endif // SDSUPPORT
 
+#ifdef PISTON_EXTRUDER
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: `#if ENABLED(PISTON_EXTRUDER)`

File: Marlin/Marlin_main.cpp [hunk 3]
```diff
@@ -4810,6 +4825,65 @@ inline void gcode_M31() {
...
+  inline void gcode_M35() { //turn on dryer fan
+    WRITE(DRYER_FAN_PIN, 1);
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: Please use `HIGH` instead of `1`.

File: Marlin/Marlin_main.cpp [hunk 4]
```diff
@@ -4810,6 +4825,65 @@ inline void gcode_M31() {
...
+    SERIAL_ECHO_START;
+    SERIAL_ECHOLN("Dryer fan ON");
+  }  
+  inline void gcode_M36() { //turn off dryer fan
+    WRITE(DRYER_FAN_PIN, 0);
```
Comment (thinkyhead) [inline Marlin/Marlin_main.cpp]: Please use `LOW` rather than `0`.

File: Marlin/pins_RAMPS.h [hunk 5]
```diff
@@ -162,7 +162,17 @@
   #define RAMPS_D10_PIN 10
 #endif
 
-#define HEATER_0_PIN     RAMPS_D10_PIN
+#ifdef PISTON_EXTRUDER
```
Comment (thinkyhead) [inline Marlin/pins_RAMPS.h]: `#if ENABLED(PISTON_EXTRUDER)`

File: Marlin/pins_RAMPS.h [hunk 6]
```diff
@@ -162,7 +162,17 @@
...
+  #define DRYER_FAN_PIN  RAMPS_D10_PIN //to control the dryer fan
+  #define HEATER_0_PIN   RAMPS_D9_PIN  //needed to pass sanity check
+  #ifdef E_ADJUST_MANUALLY
```
Comment (thinkyhead) [inline Marlin/pins_RAMPS.h]: `#if ENABLED(E_ADJUST_MANUALLY)`

File: Marlin/pins_RAMPS.h [hunk 7]
```diff
@@ -162,7 +162,17 @@
...
+    #ifndef E_ADJUST_PIN
+      #define E_ADJUST_PIN 3 //pin to read analog extruder setting
+    #endif
+  #endif
+#else
+  #define HEATER_0_PIN   RAMPS_D10_PIN
+#endif
```
Comment (thinkyhead) [inline Marlin/pins_RAMPS.h:174]: And... Actually, when defining pins for a specific function in the pins files, don't set conditions on them. We're defining the default pins that _will be used_ when the feature that uses them is enabled. When the feature is disabled, we just don't touch the pins.

File: Marlin/planner.cpp [hunk 8]
```diff
@@ -755,7 +755,12 @@ void Planner::_buffer_line(const float &a, const float &b, const float &c, const
   #endif
   if (de < 0) SBI(dm, E_AXIS);
 
-  const float esteps_float = de * volumetric_multiplier[extruder] * flow_percentage[extruder] * 0.01;
+  #if ENABLED(E_ADJUST_MANUALLY)
+    float esteps_float = de * volumetric_multiplier[extruder] * flow_percentage[extruder] * e_adjust_multiplier * 0.01;
+  #else
+    float esteps_float = de * volumetric_multiplier[extruder] * flow_percentage[extruder] * 0.01;
+  #endif
```
Comment (thinkyhead) [inline Marlin/planner.cpp]: Marlin code is leaning towards this style, to reduce repetition:
```cpp
float esteps_float = de * volumetric_multiplier[extruder] * flow_percentage[extruder] * 0.01
  #if ENABLED(E_ADJUST_MANUALLY)
    * e_adjust_multiplier
  #endif
;
```

File: Marlin/temperature.cpp [hunk 9]
```diff
@@ -1809,13 +1812,41 @@ void Temperature::isr() {
       lcd_buttons_update();
       temp_state = MeasureTemp_3;
       break;
-    case MeasureTemp_3:
-      #if HAS_TEMP_3
-        raw_temp_value[3] += ADC;
-      #endif
-      temp_state = Prepare_FILWIDTH;
-      break;
```
Comment (thinkyhead) [inline Marlin/temperature.cpp:1817]: Here I suggest keeping `MeasureTemp3` mostly as it was, but just override the new state it sets:
```cpp
temp_state =
  #if ENABLED(E_ADJUST_MANUALLY)
    Prepare_e_adjust_reading
  #else
    Prepare_FILWIDTH
  #endif
;
```

File: Marlin/temperature.cpp [hunk 10]
```diff
@@ -1809,13 +1812,41 @@ void Temperature::isr() {
...
+        break;
+    #else
+      case MeasureTemp_3:
+        #if HAS_TEMP_3
+          raw_temp_value[3] += ADC;
+        #endif
+        temp_state = Prepare_e_adjust_reading;
+        break;
 
+      case Prepare_e_adjust_reading:
```
Comment (thinkyhead) [inline Marlin/temperature.cpp:1828]: …and then you can have `#if ENABLED(E_ADJUST_MANUALLY)` just above this line ^

**Review by thinkyhead — CHANGES_REQUESTED** [review]
Looking pretty good. Patch up with the requested changes, rebase, squash, and `git push -f` the updated branch.

**Comment by BillStruve** [comment]
Made all changes. 
Did a git rebase -i HEAD~1 since all changes were made in one add/commit, did not need to squash
Did a git push -f https://github.com/BillStruve/PistonExtruder.git
Hope that works.

**Comment by BillStruve** [comment]
Does anything else need to be done?

**Comment by BillStruve** [comment]
Updated to make E adjust independent of dryer fan.

**Pull request closed without merging**